doc/developer/design/20230721_persist_fast_path.md

```diff
@@ -0,0 +1,81 @@
+- When `clusterd` receives a peek against a Persist-backed collection — as opposed to an arrangement, which is always the case today — it will create a new `PersistPeek` operation that streams the data from a Persist shard. While it’s possible to efficiently and incrementally stream consolidated data from a Persist shard, no such API exists today; we’ll need to add one.
+
+## Detailed description
+
+### Planning
+
+Today, `environmentd` will generate a fast-path plan when a dataflow can be represented as: a `Get` from an indexed collection, a map-filter-project stage, and some “finishing” logic including order-by/limit. We’ll change this to also allow the `Get` to get from a non-indexed collection backed by a Persist shard… as long as the MFP stage does not include any filters.
+
+Why no filters?
+
+- Index-based fast-path queries are often used for single-key lookups, like `select * from indexed_collection where id = 'XXX'`. Persist cannot currently support efficient queries with this shape.
+- In general, a query with a filter may need to do unbounded work to generate a finite result set size. (In the limit, we might need to scan the entire Persist shard to generate a result set of only a few rows!) If we disallow filters, we can effectively bound the amount of work we assign to `clusterd` by bounding the result set size.
+
+Today, every select query generates a peek request into an arrangement. (Even non-fast-path queries work by creating a new dataflow that writes out its data into an arrangement, then peeking into that arrangement.) This design introduces a second type of peek: one that peeks directly into the backing Persist shard. Concrentely, this means that `environmentd` will send `ComputeCommand::Peek` commands that reference a persisted collection’s `Id`, not the id of an index or arrangement.
+
+### Peeking
+
+When `clusterd` receives a peek command, it stores and tracks a `PendingPeek` struct with the arrangement and request metadata in the compute state, polls until that arrangement has data for the requested timestamp, then read that data synchronously back out of the arrangement before returning a `ComputeResponse::Peek` with the results.
+
+Since Persist’s API is asynchronous, one straightforward approach would be to fork off a new task for every incoming Peek, and hold on to the task handle. We can periodically poll the task for completion, and cancel it if the peek itself is cancelled by the user.
+
+When executing an ordinary peek, each worker reads a subset of the results out of their local arrangement… so much of the work is spread equally across all workers/processes of a replica. However, with Persist-backed peeks, we’d really only be doing the reading in *one* task running on a single node. We’ll need to take care to ensure these peek tasks are fairly distributed to avoid unduly skewing the workload.
```

Comment:
One point that might need some evaluation with this implementation approach is whether our sizing for compute replicas would still be sound. We generally now provide most (v)CPU resources to (Timely) workers, and leave only a small amount to background tasks. If these tasks become substantially more expensive, we'd need to evaluate if this setup still works.

Reply:
Fair! For interactive queries this is unlikely to be a problem, since every peek would effectively be single-threaded. But of course nothing in this doc actually proposes preventing additional concurrency... will think that over.